### PR TITLE
Fix FsDirectoryFactoryTests

### LINKED ISSUE #109681: [CI] FsDirectoryFactoryTests testStoreDirectory  and testPreload failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/7ffiubbjjlv4c/tests/:server:test/org.elasticsearch.index.store.FsDirectoryFactoryTests/testStoreDirectory

**Reproduction line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.index.store.FsDirectoryFactoryTests.testStoreDirectory" -Dtests.seed=C0F3024401134C2E -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=es-GT -Dtests.timezone=America/Grenada -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.index.store.FsDirectoryFactoryTests#testStoreDirectory`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testStoreDirectory'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.index.store.FsDirectoryFactoryTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.AssertionError: MMAPFS (MMapDirectory@/opt/buildkite-agent/builds/bk-agent-prod-gcp-1718281231524798029/elastic/elasticsearch-periodic/server/build/testrun/test/temp/org.elasticsearch.index.store.FsDirectoryFactoryTests_C0F3024401134C2E-001/tempDir-002/fooUUID/0 lockFactory=org.apache.lucene.store.NoLockFactory@5d42e600)

  at __randomizedtesting.SeedInfo.seed([C0F3024401134C2E:36E17F36229F960F]:0)
  at org.junit.Assert.fail(Assert.java:89)
  at org.junit.Assert.assertTrue(Assert.java:42)
  at org.elasticsearch.index.store.FsDirectoryFactoryTests.doTestStoreDirectory(FsDirectoryFactoryTests.java:169)
  at org.elasticsearch.index.store.FsDirectoryFactoryTests.testStoreDirectory(FsDirectoryFactoryTests.java:147)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1570)

```

### PR DESCRIPTION
**Pull request by arteam**
Unwrap the created directory to get access to original directory doing the work, so we are not affected by directory wrappers enabled by feature flags

Resolves #109681, #110210, #110211

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java [hunk 1]
```diff
@@ -156,7 +156,7 @@ private void doTestStoreDirectory(Path tempDir, String typeSettingValue, IndexMo
         Settings settings = settingsBuilder.build();
         IndexSettings indexSettings = IndexSettingsModule.newIndexSettings("foo", settings);
         FsDirectoryFactory service = new FsDirectoryFactory();
-        try (Directory directory = service.newFSDirectory(tempDir, NoLockFactory.INSTANCE, indexSettings)) {
+        try (Directory directory = unwrapDirectory(service.newFSDirectory(tempDir, NoLockFactory.INSTANCE, indexSettings))) {
```
Comment (fcofdez) [inline server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java:159]: maybe we just need to unwrap the directory in the assertions?

Comment (arteam) [reply server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java]: That makes sense. We do want to test that the wrapper behaves correctly.

File: server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java [hunk 2]
```diff
@@ -180,4 +184,16 @@ private void doTestStoreDirectory(Path tempDir, String typeSettingValue, IndexMo
...
+        var d = directory;
+        while (d != null) {
+            if (d instanceof FilterDirectory fd) {
+                d = fd.getDelegate();
+            } else {
+                break;
+            }
+        }
+        return d;
+    }
```
Comment (ywangd) [inline server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java:198]: I think this can be replaced by `FilterDirectory#unwrap`.

**Review by ywangd — APPROVED** [review]
LGTM

**Review by fcofdez — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java (hunk 2 above)
Comment (arteam) [reply server/src/test/java/org/elasticsearch/index/store/FsDirectoryFactoryTests.java]: That's nice!

**Pull request merged**